awscrt/mqtt.py

```diff
@@ -85,31 +87,100 @@ def __init__(self,
 
         super(Connection, self).__init__()
         self.client = client
-
-        def _on_connection_interrupted(error_code):
-            if on_connection_interrupted:
-                on_connection_interrupted(self, error_code)
-
-        def _on_connection_resumed(error_code, session_present):
-            if on_connection_resumed:
-                on_connection_resumed(self, error_code, session_present)
+        self._on_connection_interrupted_cb = on_connection_interrupted
+        self._on_connection_resumed_cb = on_connection_resumed
 
         self._binding = _awscrt.mqtt_client_connection_new(
+            self,
             client,
-            _on_connection_interrupted,
-            _on_connection_resumed,
         )
 
+    def _on_connection_interrupted(self, error_code):
+        if self._on_connection_interrupted_cb:
+            self._on_connection_interrupted_cb(self, error_code)
+
+    def _on_connection_resumed(self, error_code, session_present):
+        if self._on_connection_resumed_cb:
+            self._on_connection_resumed_cb(self, error_code, session_present)
+
+    def _ws_handshake_transform(self, http_request_binding, http_headers_binding, native_userdata):
+        if self._ws_handshake_transform_cb is None:
+            _awscrt.mqtt_ws_handshake_transform_complete(None, native_userdata)
+            return
+
+        def _on_complete(f):
+            _awscrt.mqtt_ws_handshake_transform_complete(f.exception(), native_userdata)
+
+        future = Future()
+        future.add_done_callback(_on_complete)
+        http_request = HttpRequest._from_bindings(http_request_binding, http_headers_binding)
+        transform_args = WebsocketHandshakeTransformArgs(self, http_request, future)
+        try:
+            self._ws_handshake_transform_cb(transform_args)
+        except Exception as e:
+            traceback.print_exc()
```

Comment:
Don't all exceptions capture a stack? What's the impetus here?

Reply:
I was trying to give the user more information, the Python exception will be swallowed as we translate this to an AWS_ERROR_HTTP_CALLBACK_FAILURE that stops the MQTT connection attempt. So I was trying to drop a breadcrumb.

But this isn't the best place to dump that information.